### PR TITLE
Collection expressions: emit Array.Empty<T>() for [] with target type T[] or IEnumerable<T>

### PR DESCRIPTION
**Pull request by cston**
Emit `Array.Empty<T>()` for an empty collection expression when the target type is:
- `T[]`
- `IEnumerable<T>`
- `IReadOnlyCollection<T>`
- `IReadOnlyList<T>`

Relates to test plan https://github.com/dotnet/roslyn/issues/66418

### CODE REVIEW

**Comment by Sergio0694** [comment]
Question: would it make sense to emit `Enumerable.Empty<T>()` instead when the target is `IEnumerable<T>`? Wouldn't that be more efficient, in that it would avoid the allocation of the enumerator (as the returned enumerable is also an enumerator)?

**Comment by CyrusNajmabadi** [comment]
Does array.empty not optimize the enumerator it returns either?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs [hunk 1]
```diff
@@ -156,11 +160,25 @@ private BoundExpression VisitCollectionInitializerCollectionExpression(BoundColl
         private BoundExpression VisitListInterfaceCollectionExpression(BoundCollectionExpression node)
         {
             Debug.Assert(!_inExpressionLambda);
-            Debug.Assert(node.Type is { });
+            Debug.Assert(node.Type is NamedTypeSymbol);
+
+            var collectionType = (NamedTypeSymbol)node.Type;
+            BoundExpression? arrayOrList;
+
+            // For [] with target type IEnumerable<T>, use Array.Empty<T>() rather than List<T>.
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:168]: could also do for IReadOnlyList/IReadOnlyCollection.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 2]
```diff
@@ -5736,13 +5897,12 @@ static MyCollection<object> F2(int x, object y)
...
-                  IL_0006:  newobj     "System.ReadOnlySpan<string>..ctor(string[])"
-                  IL_000b:  call       "MyCollection<string> MyCollectionBuilder.Create<string>(System.ReadOnlySpan<string>)"
-                  IL_0010:  ret
+                  IL_0000:  call       "string[] System.Array.Empty<string>()"
+                  IL_0005:  newobj     "System.ReadOnlySpan<string>..ctor(string[])"
+                  IL_000a:  call       "MyCollection<string> MyCollectionBuilder.Create<string>(System.ReadOnlySpan<string>)"
+                  IL_000f:  ret
                 }
                 """);
             verifier.VerifyIL("Program.F1",
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:5908]: add test where the destination is an `int*[][]`  this should be able to use `Array.Empty<int*[]>` (inner type is pointer, outer is array).

**Comment by stephentoub** [comment]
> > Does array.empty not optimize the enumerator it returns either?
> 
> It looks like `Array.Empty<T>().GetEnumerator()` returns a unique instance each time (see [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7NAExAGoAfAAQCYBGAWACgKAGAAgpoDoAZASzwCOAbkaMOATgAUjVrNYRgAKxhgMnAEowAZjFh4wMAKICcAQwA2AZ2kM5d1gEEoUUwE9OhgLYAHDK4A8HMwAfJIAlJwA4jAYhng4nrqmGNDhaDL2sk4u7l6+AUGhEdGx8YkuKVDhYWEiDGI0UhmyCsqqGtq6MPpGJhbWzfZxCUnA5jAePn6BNCHhUTHD5cmpYem2maxLo+OT+TNzxYtlSZXVtaIMQA=)).

`Array.Empty<T>()`'s GetEnumerator returns a singleton.  It does not allocate.

**Comment by alrz** [comment]
Is it recommended to use [] for immutable array as well?

**Comment by CyrusNajmabadi** [comment]
> Is it recommended to use [] for immutable array as well?

Yes.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs [hunk 3]
```diff
@@ -156,11 +160,31 @@ private BoundExpression VisitCollectionInitializerCollectionExpression(BoundColl
         private BoundExpression VisitListInterfaceCollectionExpression(BoundCollectionExpression node)
         {
             Debug.Assert(!_inExpressionLambda);
-            Debug.Assert(node.Type is { });
+            Debug.Assert(node.Type is NamedTypeSymbol);
+
+            var collectionType = (NamedTypeSymbol)node.Type;
+            BoundExpression arrayOrList;
+
+            // Use Array.Empty<T>() rather than List<T> for an empty collection expression when
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_CollectionExpression.cs:168]: I know things are in flux here but I would be fine saying this is just what we do for the empty literal. And the implementation seems fine.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 4]
```diff
@@ -5188,6 +5183,172 @@ static void Main()
...
+                        EmptyIList().Report();
+                        EmptyIReadOnlyCollection().Report();
+                        EmptyIReadOnlyList().Report();
+                    }
+                    static string[] EmptyArray() => [];
+                    static IEnumerable<string> EmptyIEnumerable() => [];
+                    static ICollection<string> EmptyICollection() => [];
+                    static IList<string> EmptyIList() => [];
+                    static IReadOnlyCollection<string> EmptyIReadOnlyCollection() => [];
+                    static IReadOnlyList<string> EmptyIReadOnlyList() => [];
```
[the remark this answers is not in the document]
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:5278]: It looks like we are calling Array.Empty for these. Did I miss something?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:5278]: Yes, these cases were added in commit 2 response to [https://github.com/dotnet/roslyn/pull/69355#discussion_r1283316793](https://github.com/dotnet/roslyn/pull/69355#discussion_r1283316793).

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 5]
```diff
@@ -1469,38 +1469,72 @@ bool isLambdaConversion(BoundExpression expr)
...
         }
 
+        private BoundExpression CreateEmptyArray(SyntaxNode syntax, ArrayTypeSymbol arrayType)
+        {
+            BoundExpression? arrayEmpty = CreateArrayEmptyCallIfAvailable(syntax, arrayType.ElementType);
+            if (arrayEmpty is { })
+            {
+                return arrayEmpty;
+            }
+            return new BoundArrayCreation(
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:1493]: nit: consider adding equivalent pseudo-code `// new ArrayType[0]`

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 6]
```diff
@@ -1469,38 +1469,72 @@ bool isLambdaConversion(BoundExpression expr)
...
+                    isDelegateCall: false,
+                    expanded: false,
+                    invokedAsExtensionMethod: false,
+                    argsToParamsOpt: default(ImmutableArray<int>),
+                    defaultArguments: default(BitVector),
+                    resultKind: LookupResultKind.Viable,
+                    type: arrayEmpty.ReturnType);
+            }
+
+            return null;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:1535]: nit: consider inverting above `if`

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 7]
```diff
@@ -2457,11 +2457,10 @@ static void Main()
             var verifier = CompileAndVerify(new[] { source, s_collectionExtensions }, expectedOutput: "[], [1, 2], [3, 4, 5], [null, 7], ");
             verifier.VerifyIL("Program.Create1", """
                 {
-                  // Code size        7 (0x7)
+                  // Code size        6 (0x6)
                   .maxstack  1
-                  IL_0000:  ldc.i4.0
-                  IL_0001:  newarr     "int"
-                  IL_0006:  ret
+                  IL_0000:  call       "int[] System.Array.Empty<int>()"
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:2462]: Do we have a test with missing `Array.Empty` API?

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 3). Minor nits only

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Pull request merged**